### DISCUSSION TITLE
Output object only an associated variable named POLLUT_CONC_0

### DISCUSSION THREAD

**Discussion by bemcdonnell** [pyswmm/pyswmm#550] [Q&A] [answered]
I would like to record the values of multiple pollutants in each step, but when I open the Output object, it doesn't seem to be able to derive the appropriate dictionary, only an associated variable named POLLUT_CONC_0. This results in me needing to traverse the Nodes(sim) object multiple times in the run.
Can the above be optimized please?

_Originally posted by @sausagechen in https://github.com/pyswmm/pyswmm/discussions/503#discussioncomment-8045299_

**Comment by bemcdonnell** [comment]
Current Workaround until this effort gets planned in:

https://github.com/pyswmm/pyswmm/discussions/514

**Comment by t298** [comment]
> 我想记录每个步骤中多个云端的值，但当我打开输出对象时，似乎无法导出适当的字典，只能导出一个名为 POLLUT_CONC_0 的关联变量。这导致我需要在运行中多次遍历 Nodes(sim) 对象。 请问上述内容可以优化吗？
> 
> _最初发布者@sausagechen在[#503 （评论）](https://github.com/pyswmm/pyswmm/discussions/503#discussioncomment-8045299)_

So how should I get the corresponding multiple pollutant concentration values according to the time and node?

![image](https://github.com/user-attachments/assets/2e212bac-2a3c-4ebc-948e-5a19b46eb26f)

**Comment by t298** [comment]
![image](https://github.com/user-attachments/assets/9a91df97-8953-4b24-9594-eadce7fd2fc2)

I tried this again, but there was no time I needed in either 'step' or 'sim' or 'node', so I had to iterate and set the time by myself?'current_time'，'report_start',''

**Comment by bemcdonnell** [comment]
You should use sim.current_time

**Comment by t298** [comment]
> You should use sim.current_time

This is the parameter in my 'inp'. The interval I get using 'sim.current_time' is 20 seconds. The interval I expect should be one day. But I set 'ROUTING_STEP 24:00:00' and got the wrong time too.

![image](https://github.com/user-attachments/assets/a30cc711-e5b2-4a60-9cd9-3d336de4974b)
![image](https://github.com/user-attachments/assets/e99283d4-5e1a-40ee-a6ed-347fbdb1f7e2)
![image](https://github.com/user-attachments/assets/081fa2f4-24c2-4c4d-a8c0-3b03a53c9a95)

**Comment by bemcdonnell** [comment]
Can you share the “options” section of your input file?

**Comment by t298** [comment]
ok，
[OPTIONS]
;;Option             Value
FLOW_UNITS           CMS
INFILTRATION         HORTON
FLOW_ROUTING         DYNWAVE
LINK_OFFSETS         DEPTH
MIN_SLOPE            0
ALLOW_PONDING        YES
SKIP_STEADY_STATE    NO

START_DATE           02/29/2004
START_TIME           00:00:00
REPORT_START_DATE    02/29/2004
REPORT_START_TIME    00:00:00
END_DATE             04/05/2004
END_TIME             00:00:00
SWEEP_START          01/01
SWEEP_END            12/31
DRY_DAYS             0
REPORT_STEP          24:00:00
WET_STEP             24:00:00
DRY_STEP             24:00:00
ROUTING_STEP         0:00:20
RULE_STEP            00:00:00

INERTIAL_DAMPING     PARTIAL
NORMAL_FLOW_LIMITED  BOTH
FORCE_MAIN_EQUATION  H-W
VARIABLE_STEP        0.75
LENGTHENING_STEP     0
MIN_SURFAREA         1.167
MAX_TRIALS           8
HEAD_TOLERANCE       0.0015
SYS_FLOW_TOL         5
LAT_FLOW_TOL         5
MINIMUM_STEP         0.5
THREADS              1

**Comment by t298** [comment]
![image](https://github.com/user-attachments/assets/eae679f1-7d9c-499e-a6c7-87bcd1daa9e5)


I tried it again, but I still had two questions:

1. There are only two pollutants in my file, but I added NodeAttribute. timers _conc_7 to 7, but he can still calculate the result？

2. With so many results, how can I know their correspondence?

**Comment by karosc** [comment, marked as answer]
SWMM uses enumerations to compute the index in the outfile from which to start parsing data. You can think of the output file as a long sequence of bytes. Various data are encoded in that sequence, some are timeseries data, some are object names, datetimes etc. 

If you arbitrarily increase the attribute enum beyond the actual number of attributes , swmm will think you want to parse an extra time series that does not exist. For example, if you have five pollutants in your model and you tell pyswmm to read the sixth (which does not exist), it will return the first attribute of the next node or link instead because that is where the sixth pollutant would be written in the output file if it did exist.

pyswmm doesn't have a good way of extending the attribute enums right now. It really depends on the user to know what they are doing. You can however use [swmm pandas](https://github.com/karosc/swmm-pandas?tab=readme-ov-file), which handles the enumeration extension for you.

`pip install swmm-pandas`

```python
import swmm.pandas as spd
out = spd.Output('./swmm_toc/out1.out')
series = out.link_series(
    attribute=('dtprsmg','b8wttak'), # pollutant names can also be found in out.link_attributes
    link='C1'  # link names can also be found out.links
)
series.plot()
```
![image](https://github.com/user-attachments/assets/69760018-5b02-4d9c-b588-a1c9a0398cc9)

I have not updated swmm-pandas since numpy 2.0 dropped, so I only guarantee it working with numpy<2 (although most of it may work with numpy 2.

`pip install numpy<2`

**Comment by t298** [comment]
@karosc  thanks
